### PR TITLE
SoftFileLock asyncio workaround (#444)

### PR DESCRIPTION
**Pull request by jw-96**
## Acknowledgment
- [X] I acknowledge that this contribution will be available under the Apache 2 license.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Summary
The SoftFileLock does not work with asyncio. This implements the workaround suggested at [#78](https://github.com/benediktschmitt/py-filelock/issues/78).

## Checklist

- [x] All tests passing
- [x] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [x] My code follows the code style of this project
(we are using `black`: you can `pip install pre-commit`,
run `pre-commit install` in the `pydra` directory
and `black` will be run automatically with each commit)

### CODE REVIEW

**Comment by satra** [comment]
@jw-96 thanks for tracking this down. to keep things similar could you instead add a contextual option

```
class PydraFileLock:

    def __init__(self, lockfile):
        self.lockfile = lockfile

    def __enter__(self):
        lock = SoftFileLock(self.lockfile)
        acquired_lock = False
        while not acquired_lock:
                lock.acquire(timeout=0)
                acquired_lock = True
            except Timeout:
                await asyncio.sleep(1)
        self.lock = lock   
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.lock.release()
        return None
```

you could add this to helpers, and then this becomes a drop in replacement for SoftFileLock wherever it is used.

**Comment by jw-96** [comment]
@satra I tried implementing your proposed change. However, I got this error:
```
SyntaxError: 'await' outside async function
```
I also tried the following: 
```
class PydraFileLock:
    """Wrapper for filelock's SoftFileLock that makes it work with asyncio."""

    def __init__(self, lockfile):
        self.lockfile = lockfile

    async def __enter__(self):
        lock = SoftFileLock(self.lockfile)
        acquired_lock = False
        while not acquired_lock:
            try:
                lock.acquire(timeout=0)
                acquired_lock = True
            except Timeout:
                await asyncio.sleep(1)
        self.lock = lock
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.lock.release()
        return None
```
However, it resulted in both tasks entering the critical section.
I also added mine and @djarecka test.

**Comment by effigies** [comment]
Is it easy to identify when we're in an async context? If so, maybe what we want is a factory:

```Python
class AsyncSoftFileLock:
    ...
    async def __enter__(...):
        ...

def get_file_lock(filename):
    if async:
        return AsyncSoftFileLock(filename)
    return SoftFileLock(filename)
```

**Comment by jw-96** [comment]
Should we use pytest-timeout to automatically stop the test after 5 seconds? Currently I am only checking if the results are correct, but if the deadlock would reoccur in the future, then the testing could get stuck.

**Comment by satra** [comment]
if this works, let's go with this. i realized there are more complications (we use softlock in both an async and a non async setting. 

i'm now curious why the change was not needed in the workflow run class which also has a softlock. perhaps we should test this out where the subtask is a workflow as well.

**Comment by mgxd** [comment]
it's been awhile since I took a look at the code, so as a refresher for myself:
- We are sync within `Task._run` (and a lock is being created there)
- We are async within `Workflow._run` (also creates a lock)

I think the idea (at the time) behind this was that `Task`s were singular commands whereas `Workflow`s needed to be expanded and waited on.

**Comment by djarecka** [comment]
`Task._run` is run in [exec_as_coro](https://github.com/nipype/pydra/blob/master/pydra/engine/workers.py#L160)

**Comment by jw-96** [comment]
@satra I was able to get it working with [__aenter__](https://www.python.org/dev/peps/pep-0492/#asynchronous-context-managers-and-async-with). I also added timeouts to the tests.

**Comment by satra** [comment]
@jw-96 - thank you. closed my PRs and just triggered this to run. the tests are taking much longer. so we may want to consider reducing the sleep time. let's see how long the tests are relative to the other branches.

**Comment by jw-96** [comment]
I noticed the same. I reduced the sleep time to 0.1 seconds and added exponential back-off capped at 4 seconds.

**Comment by satra** [comment]
>  I reduced the sleep time to 0.1 seconds and added exponential back-off capped at 4 seconds.

did these help? you haven't pushed these yet right?

**Comment by jw-96** [comment]
I only run ```pytest test_workflow.py::test_duplicate_input_on_split_wf```: 
with 1 second sleep: 11.55s
with 0.1 seconds + exp. backoff: 10.66s
I run it thrice and the run times only varied by ± 0.05s

**Comment by satra** [comment]
i'm going to merge this. let's keep the exponential backup in tow in case that comes in handy.

**Comment by satra** [comment]
actually before i merge. could you please add yourself to the .zenodo.json file?

File: pydra/engine/core.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 import cloudpickle as cp
 from filelock import SoftFileLock
+import asyncio
```
Comment (satra) [inline pydra/engine/core.py:14]: i believe this is not necessary since it is part of the default language specs now.

**Comment by jw-96** [comment]
Is the orcid necessary? I do not have one.

File: pydra/engine/core.py (hunk 1 above)
Comment (jw-96) [reply pydra/engine/core.py:14]: Yes, it seems to be not necessary

**Comment by satra** [comment]
the orcid is not necessary, but helps if you want tracking of contributions across different projects.

File: pydra/engine/helpers.py [hunk 2]
```diff
@@ -893,3 +893,29 @@ def argstr_formatting(argstr, inputs, value_updates=None):
...
+    async def __aenter__(self):
+        lock = SoftFileLock(self.lockfile)
+        acquired_lock = False
+        while not acquired_lock:
+            try:
+                lock.acquire(timeout=0)
+                acquired_lock = True
+            except Timeout:
+                await asyncio.sleep(self.timeout)
+                if self.timeout <= 2:
```
Comment (djarecka) [inline pydra/engine/helpers.py:914]: should we raise some meaningful exception here?

**Comment by jw-96** [comment]
I removed the ```import asyncio``` and added myself to the .zenodo.json. I quickly made a orcid.

File: pydra/engine/helpers.py (hunk 2 above)
Comment (satra) [reply pydra/engine/helpers.py:914]: i think this is ok. it's just used to trap. this is within a polling loop.

Comment (djarecka) [reply pydra/engine/helpers.py:914]: ok

**Comment by jw-96** [comment]
Sorry forgot a comma in zenodo file. Is there a way to only have the latest commit running?

**Comment by djarecka** [comment]
@jw-96 - don't worry. However, could I ask you for one small change? Could you please move your name before Satra. Thanks

**Comment by jw-96** [comment]
Ok, like this?

**Comment by djarecka** [comment]
yes, thanks a lot!

**Pull request merged**